Add option to manually set the number of queues for tap devices

ruocco opened the pull request:
As described in issue #9290 , it's useful sometimes to set the number of queues of a tap device to a specific values. This patchset allows doing that.

Patch 1-2 are changes in the agent, ensuring that the software fully utilized all the queues available on interface link update.

Patch 3-4 are changes in the runtime, allowing to set the number of queues via toml file or annotation

Patch 5 is a typo fix that I discovered while adding this feature.

I briefly talked to @fidencio about this in one of the Confidential Computers Community meetings.

fidencio commented:
@amshinde, mind to take a look at this one?

Apokleos commented on src/runtime/config/configuration-qemu-se.toml.in:
```diff
@@ -432,6 +432,14 @@ disable_selinux=@DEFDISABLESELINUX@
+#  > defaultMaxVCPUs will be set to default_vcpus
+#  Leaving it at default should yield the best performance in a multi-stream scenario.
+#  If your application is single-stream it's better to set it at 1.
+default_tap_queues = 0
```
Why not use the `network_queues` for net device? In runtime-rs, there is already a dedicated item `network_queues` used to set device queues in multi queues case. 

And I believe it is necessary to keep the configuration of runtime-rs/runtime-go versions as consistent as possible.

ruocco replied:
Noted. Will update the PR.

Apokleos commented on src/runtime/virtcontainers/network_linux.go:
```diff
@@ -597,7 +597,12 @@ func xConnectVMNetwork(ctx context.Context, endpoint Endpoint, h Hypervisor) err
 	caps := h.Capabilities(ctx)
 	if caps.IsMultiQueueSupported() {
-		queues = int(h.HypervisorConfig().NumVCPUs())
+		queues = int(h.HypervisorConfig().NumTapQueues)
+		if (queues <= 0) {
+			queues = int(h.HypervisorConfig().NumVCPUs())
+		} else if queues > int(h.HypervisorConfig().DefaultMaxVCPUs) {
+			queues = int(h.HypervisorConfig().DefaultMaxVCPUs)
+		}
```
I am not sure if we should provide a detection mechanism to guarantee that the network queues entered by the user are correctly linked to the number of CPUs in the VM, to ensure that the number of network queues set by the user is valid, and that the final result of the multi-queue setting meets expectations. If we do so, maybe, we should do it during the `load_config` process, or in some other way.

ruocco replied:
About the `<= 0` is to maintain behaviour, meaning that with the default value (0) it'll behave exactly as it did before, meaning that the number of queues will be set to the number of vCPUs.

About `> DefaultMaxVCPUs`, I was also not sure about this, but I saw that kata does ensure for example that DefaultVCPUs doesn't exceed the number of physical core of the system. I found similar reasoning with this limitation, meaning that setting a number of queues greater than the number of cores available to the VM it's a waste of resources without benefit. Do you think I should remove this?

Anyway, I'll refactor this at config loading time.

Apokleos reviewed (commented):
Thx @ruocco

amshinde commented on src/agent/src/random.rs:
```diff
@@ -10,6 +10,7 @@ use nix::sys::stat::Mode;
 use std::fs;
 use std::os::unix::io::{AsRawFd, FromRawFd};
 use tracing::instrument;
+use crate::util::IoctlRequestType;
```
@ruocco We typically expect an explanation for the change in the commit description. Can you add that explaning the need for this change.

amshinde commented on src/agent/src/netlink.rs:
```diff
@@ -14,6 +14,29 @@ use std::fmt;
 use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
 use std::ops::Deref;
 use std::str::{self, FromStr};
+use libc::{ioctl, ifreq, SIOCETHTOOL, __u32, c_char};
```
Same comment as previous change. 
Can you explain why this change is required?

amshinde commented on src/runtime/config/configuration-qemu-nvidia-gpu.toml.in:
```diff
@@ -467,6 +467,14 @@ disable_selinux=@DEFDISABLESELINUX@
+#  > defaultMaxVCPUs will be set to default_vcpus
+#  Leaving it at default should yield the best performance in a multi-stream scenario.
+#  If your application is single-stream it's better to set it at 1.
+default_tap_queues = 0
```
I think we made an assumption later that if the  tap queues is uninitialised (in which case it is 0) use value 1 which makes sense. 
Since we are explicitly initializing it here, would prefer `default_tap_queues = 1`.
That was it aligns with the explanation as well.

amshinde replied:
Also curious about the kind of workload you were running, where you saw better performance with single queue.

ruocco replied:
Maybe the explanation I wrote isn't explanatory enough :)

When uninitialized (or zero) it will be set to the same value it was before the change, i.e. equal to the number of `default_vcpus`. If set at one it's going to be single-queue, which is probably not what you want in most cases.

ruocco replied:
>  Also curious about the kind of workload you were running, where you saw better performance with single queue.

It's an Optical Proximity Correction software (calibre by Siemens)

ruocco commented:
@Apokleos @amshinde 

I've updated the PR with the requested changes:

- The configuration is now called `network_queues`
- Added commit messages
- The sanity check is at config time.

However, now that the sanity check is earlier in the process there are some issues: when the config file has `default_vcpus = N` and we override it with the annotation `io.katacontainers.config.hypervisor.default_vcpus = M`, without setting the number of `network_queues` in its own annotation, they will be set at `N` instead of `M`.
If we were to change the code so that the `default_vcpus` annotation override would also set the `network_queues` when that particular annotation is missing, that would also override the user-set value in the config file. Which is probably not the expected behaviour.

Also there's conflict on the `network_queues` sanity check at annotation time, since there's no way of knowing which are the correct values for default_vcpus and default_maxvcpus to check against, given that they might have been overridden from other annotations. 


Given how the `network_queues` and `default_vcpus` are tied I would come back to the original proposal, i.e. a single check done when creating the tap device.

Apokleos commented:
> @Apokleos @amshinde
> 
> I've updated the PR with the requested changes:
> 
> * The configuration is now called `network_queues`
> * Added commit messages
> * The sanity check is at config time.
> 
> However, now that the sanity check is earlier in the process there are some issues: when the config file has `default_vcpus = N` and we override it with the annotation `io.katacontainers.config.hypervisor.default_vcpus = M`, without setting the number of `network_queues` in its own annotation, they will be set at `N` instead of `M`. If we were to change the code so that the `default_vcpus` annotation override would also set the `network_queues` when that particular annotation is missing, that would also override the user-set value in the config file. Which is probably not the expected behaviour.
> 
> Also there's conflict on the `network_queues` sanity check at annotation time, since there's no way of knowing which are the correct values for default_vcpus and default_maxvcpus to check against, given that they might have been overridden from other annotations.
>
The `network_queues` set MUST be no more than the current vCPUs of kata pod(vm and containers), right ?

Apokleos commented on src/runtime/virtcontainers/network_linux.go:
```diff
@@ -597,7 +597,7 @@ func xConnectVMNetwork(ctx context.Context, endpoint Endpoint, h Hypervisor) err
 	caps := h.Capabilities(ctx)
 	if caps.IsMultiQueueSupported() {
-		queues = int(h.HypervisorConfig().NumVCPUs())
+		queues = int(h.HypervisorConfig().NetworkQueues)
```
Here, the original queues which comes from `int(h.HypervisorConfig().NumVCPUs())` is equal to the current vcpus. 
If we want to check the validation of network queues setting , maybe it is a choice, or there's a better one.

Apokleos reviewed (commented):
Thx for your updating the code.

ruocco commented:
> > @Apokleos @amshinde
> > I've updated the PR with the requested changes:
> > 
> > * The configuration is now called `network_queues`
> > * Added commit messages
> > * The sanity check is at config time.
> > 
> > However, now that the sanity check is earlier in the process there are some issues: when the config file has `default_vcpus = N` and we override it with the annotation `io.katacontainers.config.hypervisor.default_vcpus = M`, without setting the number of `network_queues` in its own annotation, they will be set at `N` instead of `M`. If we were to change the code so that the `default_vcpus` annotation override would also set the `network_queues` when that particular annotation is missing, that would also override the user-set value in the config file. Which is probably not the expected behaviour.
> > Also there's conflict on the `network_queues` sanity check at annotation time, since there's no way of knowing which are the correct values for default_vcpus and default_maxvcpus to check against, given that they might have been overridden from other annotations.
> 
> The `network_queues` set MUST be no more than the current vCPUs of kata pod(vm and containers), right ?

Well it's not a *hard* requirement, you can have it be greater than that, it's just that you won't have any benefits with this, only downsides. Same concept as having default vCPUs greater than physical CPUs, you *can* do it, it's just pointless, which is why it's not allowed in kata.

ruocco commented on src/agent/src/netlink.rs:
```diff
@@ -130,7 +153,41 @@ impl Handle {
+			SockFlag::empty(),
+			None,
+			);
+		let fd = sock.unwrap();
```
Ah, probably not, I'll swap this with the `?` operator on my next PR update when we decide the best way on moving forward with the validity check of the number of queues and the default value when the value is not set
